### PR TITLE
WIP: Feature/dynamic runner types + endpoint security

### PR DESCRIPTION
**Pull request by quvide**
TODO

- [x] Beautiful frontend (wip by @sofiatengvall)
- [x] Pass tests
- [x] Write docs
- [x] Postgres migrations (wip by @quvide)
- [x] Polling build status
- [x] Refetch containers after delete
- [x] Fix runner ownership check
- [ ] See issues in slack thread

### CODE REVIEW

File: client/src/create-runner.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import React from 'react'
+import R from 'ramda'
```
Comment (bqqbarbhg) [inline client/src/create-runner.js]: Use `import * as R from 'ramda'` because JS

File: docs/server-config.md [hunk 2]
```diff
@@ -10,3 +10,4 @@ environment variables. You can specify the configuration file to use with the `-
 | --- | --- | --- |
 | `port` | `Integer` | Port to bind the API server to, overridden by env `APP_SERVER_PORT`. |
 | `dev` | `Boolean` | If set to true the server runs in development mode serving also static content. |
+| `dockerWindows` | `Boolean` | Set this to true if you're using Docker for Windows (Only Win10 Pro/Education). If you're using Docker Toolbox (other Windows versions), leave this to false. Default is false. |
```
Comment (bqqbarbhg) [inline docs/server-config.md]: Could this be autodetected?

File: server/src/database.js [hunk 3]
```diff
@@ -102,3 +105,69 @@ export async function verifyUser(user, secret) {
...
+export async function confirmContainerOwnership(name, users) {
+  console.log(name, users)
+  const rows = await db.query('SELECT user_id FROM containers WHERE subdomain=? LIMIT 1', [name])
+  if (!rows || rows.length === 0) {
+    return false
+  }
+
+  const ownerUserId = rows[0].user_id
+  if (users.hasOwnProperty(ownerUserId)) {
+    return ownerUserId
```
Comment (bqqbarbhg) [inline server/src/database.js]: Is this multi-type return necessary, can you for example throw an error here?

File: server/src/database/migrations/sqlite/001-setup.sql [hunk 4]
```diff
@@ -70,5 +70,16 @@ CREATE TABLE reactions (
 -- Table: Migrations
 CREATE TABLE migrations (id INTEGER PRIMARY KEY, file VARCHAR(255) UNIQUE);
 
+-- Table: Instance runners
+CREATE TABLE instance_runners (
+  tag       VARCHAR(32) UNIQUE NOT NULL,
+  time     VARCHAR(30) DEFAULT (CURRENT_TIMESTAMP) NOT NULL,
+  name     VARCHAR(32) NOT NULL,
+  user_id  CHAR(8) NOT NULL,
+  size     INT,
```
Comment (bqqbarbhg) [inline server/src/database/migrations/sqlite/001-setup.sql:96]: Can this be NULL?

File: server/src/docker.js [hunk 5]
```diff
@@ -1,8 +1,17 @@
 import Docker from 'dockerode'
+import fs from 'fs'
+import os from 'os'
+import R from 'ramda'
```
Comment (bqqbarbhg) [inline server/src/docker.js]: `* as R` again

File: server/src/database/migrations/sqlite/001-setup.sql (hunk 4 above)
Comment (quvide) [reply server/src/database/migrations/sqlite/001-setup.sql:96]: This is intended, it's null on images with pending status, but it actually never gets updated in

Comment (quvide) [reply server/src/database/migrations/sqlite/001-setup.sql]: and I just realized that default images are not in the db anymore...

File: server/src/docker.js (hunk 5 above)
Comment (quvide) [reply server/src/docker.js]: fixed!

File: client/src/create-runner.js (hunk 1 above)
Comment (quvide) [reply client/src/create-runner.js]: fixed!

File: docs/server-config.md (hunk 2 above)
Comment (quvide) [reply docs/server-config.md]: Maybe, but let's do it in some other PR preferrably

File: server/src/database.js (hunk 3 above)
Comment (quvide) [reply server/src/database.js]: Not necessary, this should be discussed

File: server/src/database/migrations/sqlite/001-setup.sql (hunk 4 above)
Comment (quvide) [reply server/src/database/migrations/sqlite/001-setup.sql:96]: This was fixed in c216773

File: server/src/database.js (hunk 3 above)
Comment (quvide) [reply server/src/database.js]: I changed the behaviour to throw in 7820db4

File: client/src/scss/atoms/_toast.scss [hunk 6]
```diff
@@ -0,0 +1,4 @@
+
+$rt-z-index: 999999999999;
```
Comment (HenrikHelenius) [inline client/src/scss/atoms/_toast.scss]: refactor to sensible, use z-index map from variables

File: client/src/scss/views/create-runner.scss [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  &.table-header {
+    font-weight: 600;
+  }
+
+  &.default {
+    background: $palette-lighter;
+  }
+}
+
+.cell {
```
Comment (HenrikHelenius) [inline client/src/scss/views/create-runner.scss]: rename from cell to some more descriptive when not having table layout

File: client/src/scss/views/dashboard-view.scss [hunk 8]
```diff
@@ -36,6 +36,18 @@
...
 }
 
+.top-button-section {
+  display: flex;
+  align-items: center;
+  justify-content: right;
+
+  margin: 0 -$spacer/2;
+
+  &>* {
```
Comment (HenrikHelenius) [inline client/src/scss/views/dashboard-view.scss]: spaces

File: client/src/scss/views/dashboard-view.scss [hunk 9]
```diff
@@ -36,6 +36,18 @@
   box-shadow: 0 0 10px $palette-shadow-light;
 }
 
+.top-button-section {
```
Comment (HenrikHelenius) [inline client/src/scss/views/dashboard-view.scss]: is this duplicated somewhere?

File: client/src/scss/views/dashboard-view.scss [hunk 10]
```diff
@@ -36,6 +36,18 @@
...
+  display: flex;
+  align-items: center;
+  justify-content: right;
+
+  margin: 0 -$spacer/2;
+
+  &>* {
+    margin: 0 $spacer/2;
```
Comment (HenrikHelenius) [inline client/src/scss/views/dashboard-view.scss]: space

File: client/src/index.js [hunk 11]
```diff
@@ -146,6 +148,17 @@ class MainView extends React.Component {
           taggedElementXPath={this.state.taggedElementXPath}
           unsetTaggedElement={this.unsetTaggedElement}
         />
+        <ToastContainer
+          position="bottom-center"
+          autoClose={5000}
+          hideProgressBar={false}
+          newestOnTop
+          closeOnClick
+          rtl={false}
```
Comment (HenrikHelenius) [inline client/src/index.js:187]: check that you are not declaring default values, i think this is false by default?

File: client/src/scss/views/create-runner.scss [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+.row {
+  display: flex;
+  flex-flow: row;
+  width: 100%;
+  box-shadow: 0 2px 10px $palette-shadow-default;
+  margin-bottom: $baseline;
+
+  &.table-header {
```
Comment (HenrikHelenius) [inline client/src/scss/views/create-runner.scss]: is it table?

File: client/src/site.js [hunk 13]
```diff
@@ -18,6 +20,8 @@ import { setUsers } from './globals'
 import apiCall from './api-call'
 // Styles
 import styles from './scss/_base.scss'
+import './scss/atoms/_toast.scss'
+import CreateRunner from './create-runner'
```
Comment (HenrikHelenius) [inline client/src/site.js]: reorder components with components and so on

File: client/src/site.js [hunk 14]
```diff
@@ -84,17 +88,34 @@ const initialize = () => {
...
-      </Router>
-    </Provider>,
+    <div className={css('management-container')}>
+      <ToastContainer
+        position="bottom-center"
+        autoClose={5000}
+        hideProgressBar={false}
+        newestOnTop
+        closeOnClick
+        rtl={false}
```
Comment (HenrikHelenius) [inline client/src/site.js:103]: same as in other place, default values

File: client/src/create-runner.js [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    })
+  }
+
+  render() {
+    return (
+      <div className={css('runner-view-container')}>
+        <div className={css('inner-container')}>
+          <div className={css('default-runner-container')}>
+            <h2>Default runners</h2>
+            {this.state.defaultRunners.map(runner => (<DefaultRunnerRow
```
Comment (HenrikHelenius) [inline client/src/create-runner.js]: newline

File: client/src/create-runner.js [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+              key={runner.tag}
+              runner={runner}
+            />
+            )) }
```
Comment (HenrikHelenius) [inline client/src/create-runner.js]: newline

File: client/src/create-runner.js [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+          <div className={css('runner-table')}>
+            <h2>Your runners</h2>
+            <div className={css('table-header', 'row')}>
+              <div className={css('cell')}>Tag</div>
+              <div className={css('cell')}>Created</div>
+              <div className={css('cell')}>Status</div>
+              <div className={css('cell')} />
+            </div>
+            <div className={css('table-body')}>
+              {this.state.instanceRunners.map(runner => (<RunnerRow
```
Comment (HenrikHelenius) [inline client/src/create-runner.js]: newline

File: client/src/create-runner.js [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+                key={runner.tag}
+                runner={runner}
+                deleteRunnerCallback={this.deleteRunner}
+              />
+              )) }
```
Comment (HenrikHelenius) [inline client/src/create-runner.js]: newline

File: client/src/create-runner.js [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+    <div className={css('row')}>
+      <span className={css('cell')}>{runner.tag}</span>
+      <div className={css('cell')}>
+        <Moment
+          className={css('timestamp')}
+          date={runner.time}
+          format="DD.MM.YYYY HH:mm"
+        />
+      </div>
+      <div className={runner.status === 'fail' ? css('cell', 'status-fail') : css('cell')}>
```
Comment (HenrikHelenius) [inline client/src/create-runner.js]: use css({'true-class': variable}) syntax

File: client/src/create-runner.js [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+      </div>
+    </div>
+  )
+}
+
+function DefaultRunnerRow(props) {
+  const { runner } = props
+  return (
+    <div className={css('default', 'row')}>
+      <span className={css('cell')}>{runner.tag}</span>
```
Comment (HenrikHelenius) [inline client/src/create-runner.js]: newline

**Review by HenrikHelenius — CHANGES_REQUESTED** [review]

File: client/package.json [hunk 21]
```diff
@@ -58,7 +58,9 @@
     "webpack-cli": "^3.1.2",
     "whatwg-fetch": "^3.0.0"
   },
-  "dependencies": {},
+  "dependencies": {
+    "react-toastify": "^4.5.2"
```
Comment (tobbe-j) [inline client/package.json]: Move to devDeps as with all other deps

File: server/src/database/migrations/postgres/004-instance-runners.sql [hunk 22]
```diff
@@ -0,0 +1,14 @@
+BEGIN TRANSACTION;
```
Comment (tobbe-j) [inline server/src/database/migrations/postgres/004-instance-runners.sql]: Remember to check that migration number does not clash with other new migrations from different branches

File: client/src/scss/views/dashboard-view.scss (hunk 8 above)
Comment (quvide) [reply client/src/scss/views/dashboard-view.scss]: doesn't exist anymore

File: client/src/scss/views/dashboard-view.scss (hunk 9 above)
Comment (quvide) [reply client/src/scss/views/dashboard-view.scss]: doesn't exist anymore

File: client/src/scss/views/dashboard-view.scss (hunk 10 above)
Comment (quvide) [reply client/src/scss/views/dashboard-view.scss]: doesn't exist anymore

File: client/src/create-runner.js (hunk 16 above)
Comment (quvide) [reply client/src/create-runner.js]: fixed

File: client/src/create-runner.js (hunk 18 above)
Comment (quvide) [reply client/src/create-runner.js]: fixed

File: client/package.json (hunk 21 above)
Comment (quvide) [reply client/package.json]: fixed

File: client/src/create-runner.js (hunk 20 above)
Comment (quvide) [reply client/src/create-runner.js]: fixed

File: client/src/create-runner.js (hunk 17 above)
Comment (quvide) [reply client/src/create-runner.js]: fixed

**Review by kalomak — APPROVED** [review]
No legacy code 👌🏿

**Review by HenrikHelenius — APPROVED** [review]

**Review by tobbe-j — APPROVED** [review]
Ship it

**Pull request merged**